src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -2835,6 +2835,76 @@ public override IOperation VisitCoalesce(ICoalesceOperation operation, int? capt
             return result;
         }
 
+        public override IOperation VisitCoalesceAssignment(ICoalesceAssignmentOperation operation, int? captureIdForResult)
+        {
+            SpillEvalStack();
+
+            // If we're in a statement context, we elide the capture of the result of the assignment, as it will
+            // just be wrapped in an expression statement that isn't used anywhere and isn't observed by anything.
+            bool isStatement = operation.Parent.Kind == OperationKind.ExpressionStatement;
+            Debug.Assert(captureIdForResult == null || !isStatement);
+
+            RegionBuilder resultCaptureRegion = _currentRegion;
+
+            EvalStackFrame frame = PushStackFrame();
+
+            IOperation testExpression;
+
+            PushOperand(Visit(operation.Target));
+            SpillEvalStack();
+            testExpression = PopOperand();
+
+            var whenNull = new BasicBlockBuilder(BasicBlockKind.Block);
+            var afterCoalesce = new BasicBlockBuilder(BasicBlockKind.Block);
+
+            int resultCaptureId = isStatement ? -1 : captureIdForResult ?? GetNextCaptureId(resultCaptureRegion);
+
+            LinkBlocks(CurrentBasicBlock,
+                       Operation.SetParentOperation(MakeIsNullOperation(testExpression), null),
+                       jumpIfTrue: true,
+                       RegularBranch(whenNull));
+
+            if (!isStatement)
+            {
+                _currentBasicBlock = null;
+
+                Debug.Assert(testExpression.Kind == OperationKind.FlowCaptureReference);
+                AddStatement(new FlowCapture(resultCaptureId, operation.Syntax, OperationCloner.CloneOperation(testExpression)));
```

Comment:
>OperationCloner.CloneOperation(testExpression)) [](start = 80, length = 47)

I think this is going to be interpreted as another read from the Target, is this going to accurately reflect what is happening?

Reply:
We spill the eval stack, so this is always a capture of a capture. I don't think that that implies another read, since it's a captured value.

---
In reply to: [211441273](https://github.com/dotnet/roslyn/pull/29063#discussion_r211441273) [](ancestors = 211441273)

Reply:
> We spill the eval stack, so this is always a capture of a capture. I don't think that that implies another read, since it's a captured value.

Is this really a value though? How can we assign to it below? Do we have similar pattern in other situations?

---
In reply to: [211450764](https://github.com/dotnet/roslyn/pull/29063#discussion_r211450764) [](ancestors = 211450764,211441273)

Reply:
After offline discussion, I'm going to take the approach of separate captures for the "location" and "value" of the left hand side. This is the same approach that the local rewriter takes, and while we don't have the concept of location vs value in captures today, this is something we want to add and this will be compatible when we do.

---
In reply to: [211684153](https://github.com/dotnet/roslyn/pull/29063#discussion_r211684153) [](ancestors = 211684153,211450764,211441273)